File: midonet-api/src/main/java/org/midonet/api/neutron/NeutronResource.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.api.neutron;

import java.net.URI;

import javax.annotation.security.RolesAllowed;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.SecurityContext;
import javax.ws.rs.core.UriInfo;

import com.google.inject.Inject;

import org.midonet.cluster.auth.AuthRole;
import org.midonet.api.neutron.loadbalancer.LBResource;
import org.midonet.api.neutron.loadbalancer.LBUriBuilder;
import org.midonet.api.rest_api.AbstractResource;
import org.midonet.api.rest_api.RestApiConfig;
import org.midonet.cluster.rest_api.neutron.NeutronMediaType;
import org.midonet.cluster.rest_api.neutron.NeutronUriBuilder;
import org.midonet.cluster.rest_api.neutron.models.Neutron;

@Path(NeutronUriBuilder.NEUTRON)
public class NeutronResource extends AbstractResource {

    private final NeutronResourceFactory factory;

    @Inject
    public NeutronResource(RestApiConfig config, UriInfo uriInfo,
                           SecurityContext context,
                           NeutronResourceFactory factory) {
        super(config, uriInfo, context, null, null);
        this.factory = factory;
    }

    @Path(NeutronUriBuilder.NETWORKS)
    public NetworkResource getNetworkResource() {
        return factory.getNeutronNetworkResource();
    }

    @Path(NeutronUriBuilder.SUBNETS)
    public SubnetResource getSubnetResource() {
        return factory.getNeutronSubnetResource();
    }

    @Path(NeutronUriBuilder.PORTS)
    public PortResource getPortResource() {
        return factory.getNeutronPortResource();
    }

    @Path(NeutronUriBuilder.ROUTERS)
    public RouterResource getRouterResource() {
        return factory.getNeutronRouterResource();
    }

    @Path(NeutronUriBuilder.FLOATING_IPS)
    public FloatingIpResource getFloatingIpResource() {
        return factory.getNeutronFloatingIpResource();
    }

    @Path(NeutronUriBuilder.SECURITY_GROUPS)
    public SecurityGroupResource getSecurityGroupResource() {
        return factory.getNeutronSecurityGroupResource();
    }

    @Path(NeutronUriBuilder.SECURITY_GROUP_RULES)
    public SecurityGroupRuleResource getSecurityGroupRuleResource() {
        return factory.getNeutronSecurityGroupRuleResource();
    }

    @Path(LBUriBuilder.LB)
    public LBResource getLoadBalancerResource() {
        return factory.getLoadBalancerResource();
    }

    /**
     * Handler to getting a neutron object.
     *
     * @return A Neutron object.
     */
    @GET
    @RolesAllowed(AuthRole.ADMIN)
    @Produces({NeutronMediaType.NEUTRON_JSON_V1,
               NeutronMediaType.NEUTRON_JSON_V2})
    public Neutron get() {

        Neutron neutron = new Neutron();

        URI baseUri = getBaseUri();
        neutron.uri = NeutronUriBuilder.getNeutron(baseUri);
        neutron.networks = NeutronUriBuilder.getNetworks(baseUri);
        neutron.networkTemplate = NeutronUriBuilder.getNetworkTemplate(
                baseUri);
        neutron.subnets = NeutronUriBuilder.getSubnets(baseUri);
        neutron.subnetTemplate = NeutronUriBuilder.getSubnetTemplate(baseUri);
        neutron.ports = NeutronUriBuilder.getPorts(baseUri);
        neutron.portTemplate = NeutronUriBuilder.getPortTemplate(baseUri);
        neutron.routers = NeutronUriBuilder.getRouters(baseUri);
        neutron.routerTemplate = NeutronUriBuilder.getRouterTemplate(baseUri);
        neutron.addRouterInterfaceTemplate =
                NeutronUriBuilder.getAddRouterInterfaceTemplate(baseUri);
        neutron.removeRouterInterfaceTemplate =
                NeutronUriBuilder.getRemoveRouterInterfaceTemplate(baseUri);
        neutron.floatingIps = NeutronUriBuilder.getFloatingIps(baseUri);
        neutron.floatingIpTemplate = NeutronUriBuilder.getFloatingIpTemplate(
                baseUri);
        neutron.securityGroups = NeutronUriBuilder.getSecurityGroups(baseUri);
        neutron.securityGroupTemplate =
                NeutronUriBuilder.getSecurityGroupTemplate(baseUri);
        neutron.securityGroupRules =
                NeutronUriBuilder.getSecurityGroupRules(baseUri);
        neutron.securityGroupRuleTemplate =
                NeutronUriBuilder.getSecurityGroupRuleTemplate(baseUri);
        neutron.loadBalancer = LBResource.buildLoadBalancer(baseUri);
        return neutron;
    }
}

File: cluster/midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/neutron/resources/NeutronResource.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.neutron.resources

import java.net.URI
import javax.ws.rs.core.UriInfo
import javax.ws.rs.{GET, Path, Produces}

import com.google.inject.Inject
import org.midonet.cluster.data.neutron.models.Neutron
import org.midonet.cluster.rest_api.neutron.NeutronResourceUris._
import org.midonet.cluster.rest_api.neutron.models.{Neutron, LoadBalancer}
import org.midonet.cluster.services.rest_api.neutron.NeutronMediaType
import org.midonet.cluster.services.rest_api.neutron.plugin.NeutronZoomPlugin

@Path(NEUTRON) class NeutronResource @Inject() (uriInfo: UriInfo,
                                                zoomPlugin: NeutronZoomPlugin) {

    private def buildLoadBalancer(baseUri: URI): LoadBalancer = {
        val lb: LoadBalancer = new LoadBalancer
        lb.healthMonitors = getUri(baseUri, HEALTH_MONITORS)
        lb.healthMonitorTemplate = getUri(baseUri, HEALTH_MONITORS, ID_TOKEN)
            .toString
        lb.members = getUri(baseUri, MEMBERS)
        lb.memberTemplate = getUri(baseUri, MEMBERS, ID_TOKEN).toString
        lb.pools = getUri(baseUri, POOLS)
        lb.poolTemplate = getUri(baseUri, POOLS, ID_TOKEN).toString
        lb.vips = getUri(baseUri, VIPS)
        lb.vipTemplate = getUri(baseUri, VIPS, ID_TOKEN).toString
        lb.poolHealthMonitor = getUri(baseUri, HEALTH_MONITORS)
        lb
    }

    @Path(NETWORKS)
    def getNetworkResource: NetworkResource = {
        new NetworkResource(uriInfo, zoomPlugin)
    }

    @Path(SUBNETS)
    def getSubnetResource: SubnetResource = {
        new SubnetResource(uriInfo, zoomPlugin)
    }

    @Path(PORTS)
    def getPortResource: PortResource = {
        new PortResource(uriInfo, zoomPlugin)
    }

    @Path(ROUTERS)
    def getRouterResource: RouterResource = {
        new RouterResource(uriInfo, zoomPlugin)
    }

    @Path(FLOATING_IPS)
    def getFloatingIpResource: FloatingIpResource = {
        new FloatingIpResource(uriInfo, zoomPlugin)
    }

    @Path(SECURITY_GROUPS)
    def getSecurityGroupResource: SecurityGroupResource = {
        new SecurityGroupResource(uriInfo, zoomPlugin)
    }

    @Path(SECURITY_GROUP_RULES)
    def getSecurityGroupRuleResource: SecurityGroupRuleResource = {
        new SecurityGroupRuleResource(uriInfo, zoomPlugin)
    }

    @Path(LB) def getLoadBalancerResource: LBResource = {
        new LBResource(uriInfo, zoomPlugin)
    }

    @GET
    @Produces(Array(NeutronMediaType.NEUTRON_JSON_V1,
                    NeutronMediaType.NEUTRON_JSON_V2))
    def get: Neutron = {
        val neutron: Neutron = new Neutron
        val baseUri: URI = uriInfo.getBaseUri
        neutron.uri = getUri(baseUri, NEUTRON)
        neutron.networks = getUri(baseUri, NETWORKS)
        neutron.networkTemplate = getUri(baseUri, NETWORKS, ID_TOKEN).toString
        neutron.subnets = getUri(baseUri, SUBNETS)
        neutron.subnetTemplate = getUri(baseUri, SUBNETS, ID_TOKEN).toString
        neutron.ports = getUri(baseUri, PORTS)
        neutron.portTemplate = getUri(baseUri, PORTS, ID_TOKEN).toString
        neutron.routers = getUri(baseUri, ROUTERS)
        neutron.routerTemplate = getUri(baseUri, ROUTERS, ID_TOKEN).toString
        neutron.floatingIps = getUri(baseUri, FLOATING_IPS)
        neutron.floatingIpTemplate = getUri(baseUri, FLOATING_IPS, ID_TOKEN)
            .toString
        neutron.securityGroups = getUri(baseUri, SECURITY_GROUPS)
        neutron.securityGroupTemplate = getUri(baseUri, SECURITY_GROUPS,
                                               ID_TOKEN).toString
        neutron.securityGroupRules = getUri(baseUri, SECURITY_GROUP_RULES)
        neutron.securityGroupRuleTemplate = getUri(baseUri,
                                                   SECURITY_GROUP_RULES,
                                                   ID_TOKEN).toString
        neutron.loadBalancer = buildLoadBalancer(baseUri)
        neutron.removeRouterInterfaceTemplate = neutron.routerTemplate + "/" +
                                                REMOVE_ROUTER_INTF
        neutron.addRouterInterfaceTemplate = neutron.routerTemplate + "/" +
                                             ADD_ROUTER_INTF
        neutron
    }
}